File: Mjlogreader.java
import java.io.*;
import java.nio.*;
import java.util.*;
/**
 * Reads the mjlog and outputs mahjong data. Only takes the player 3's data for now.
 * @author Kochi Nakajima
 */

/** Hais array
 0 ~ 8 Manzu 1,2,3.....9
 9 ~ 17 Pinzu
 18 ~ 26 Souzu
 27 ~ 30 Directions E, S, W, N
 31 ~ 33 Dragons W, G, R
 test
*/

public class Mjlogreader{
  int[] hand;
  int[] discardPile;
  int discarded;
  Integer PLAYERID = 0;
  String data;
  BufferedReader br;
  String TSUMO;
  String DAHAI;
  final static String ENDSIGN = "/>";
  String[] temp;
  int[] startGame;
  int junme;
  int curr_game;
  int curr_ptr;

  public Mjlogreader(String filename){
    hand = new int[34];
    discardPile = new int[34];
    startGame = new int[20];
    switchPlayer();
    try{
      br = new BufferedReader(new FileReader(filename));
      data = br.readLine();
      if(PLAYERID != 3){
        temp = data.substring(data.indexOf("hai" + PLAYERID.toString() +  "=") + 6, data.indexOf("hai" + (PLAYERID + 1) + "=") - 2).split(",");
      }else{
        temp = data.substring(data.indexOf("hai" + PLAYERID.toString() +  "=") + 6, data.indexOf(ENDSIGN, data.indexOf("hai" + PLAYERID.toString() +  "=")) - 1).split(",");
      }
      for(int i = 0; i < temp.length; i++){
        hand[tileConverter(Integer.parseInt(temp[i]))]++;
      }
      int idx = 0;
      startGame[idx] = data.indexOf("<INIT");
      while(data.indexOf("<INIT", startGame[idx] + 1) != -1){
        idx++;
        startGame[idx] = data.indexOf("<INIT", startGame[idx - 1] + 1);
      }
      junme = 0;
    }catch(IOException e){
      System.out.println("error");
    }
  }

  private void switchPlayer(){
    switch(PLAYERID){
      case 0:
        TSUMO = "<T";
        DAHAI = "<D";
        break;
      case 1:
        TSUMO = "<U";
        DAHAI = "<E";
        break;
      case 2:
        TSUMO = "<V";
        DAHAI = "<F";
        break;
      case 3:
        TSUMO = "<W";
        DAHAI = "<G";
        break;
    }
  }


  public int tileConverter(int mjlognumber){
    return mjlognumber >> 2;
  }


  public int[] getHand(){
    return hand;
  }

  public int[] getDiscardPile(){
    return discardPile;
  }

  public int getDiscardedTile(){
    return discarded;
  }

  /**
   * Goes around all the players
   * For naki, just come back.
   * When junme == 0, it will only TSUMO
   * When junme > 0, it will discard and Tsumo.
   */
  public boolean next(){
    int tsumo;
    if(junme == 0){
      junme++;
      curr_ptr = startGame[curr_game];
      tsumo = getNextTile(true);
      System.out.println("Tsumo: " + tsumo);
      hand[tsumo]++;
    }else if(curr_ptr > startGame[curr_game + 1] || curr_ptr == -1){
      System.out.println("--------");
      System.out.println("End of Kyoku");
      System.out.println("--------");
      curr_game++;
      if(startGame[curr_game] == 0){
        System.out.println("--------");
        System.out.println("--------");
        System.out.println("END OF THE GAME");
        return false;
      }else{
        System.out.println("Press enter to continue");
        Scanner s = new Scanner(System.in);
        if(s.nextLine() != null){
          newGame();
        }
      }
    }else{
      junme++;
      discarded = getNextTile(false);
      System.out.println("Discard: " + discarded);
      System.out.println("Junme: " + junme);
      hand[discarded]--;
      discardPile[discarded]++;
      tsumo = getNextTile(true);
      System.out.println("Tsumo: " + tsumo);
      hand[tsumo]++;
    }
    return true;
  }

  private void newGame(){
    temp = data.substring(data.indexOf("hai2=", startGame[curr_game]) + 6, data.indexOf("hai3=", startGame[curr_game]) - 2).split(",");
    Arrays.fill(hand, 0);
    Arrays.fill(discardPile, 0);
    for(int i = 0; i < temp.length; i++){
      hand[tileConverter(Integer.parseInt(temp[i]))]++;
    }
    junme = 0;
  }

  private void printHandArray(){
    System.out.println("---------");

    for(int i = 0; i < 34; i++) {
    			if(i == 0) System.out.print("M: ");
    			if(i == 9) System.out.print("\nP: ");
    			if(i == 18) System.out.print("\nS: ");
    			if(i == 27) System.out.print("\nW: ");
    			if(i == 31) System.out.print("\nD: ");
    			System.out.print(hand[i] + ", ");
    		}
    	System.out.println("\n---------");
  }

  private int getNextTile(boolean isTsumo){
    String indicator;
    indicator = (isTsumo)? TSUMO : DAHAI;
    curr_ptr = data.indexOf(indicator, curr_ptr);
    int returnVal = tileConverter(Integer.parseInt(data.substring(data.indexOf(indicator, curr_ptr) + 2, data.indexOf(ENDSIGN, curr_ptr))));
    return returnVal;
  }
  
  private void toCSV(String filename){
    //Need shantensuu
    //Write out in format of Junme, Shanten, Tehai, Sutehai

    String data = "dataToWrite";
    try{
      FileWriter fw = new FileWriter(filename + ".csv");
      fw.append(data);
      fw.flush();
      fw.close();
    }catch(IOException e){
      e.printStackTrace();
    }
  }

  public static void main(String[] args){
    Mjlogreader mr = new Mjlogreader("sample2.mjlog");
    do{
      mr.printHandArray();
    }while(mr.next());
    mr.toCSV("test");
  }
  
  

}

File: Mjlogreader.py
# Get Tehai from hai0, hai1, hai2, hai3
#Use T, U, V, W for Tsumos
#Use D, E, F, G for Dahai
#Keep track of hand and Shanten
#Stack up each Junme in specific format
#Junme, Tehai, Shanten, Sutehai

"""
0 ~ 8 Manzu
9 ~ 17 Pinzu
18 ~ 26 Souzu
27 ~ 30 Directions E, S, W, N
31 ~ 33 Dragons W, G, R

In mjlog,
0 ~ 35 Manzu
36 ~ 71 Souzu
72 ~ 107 Souzu
108 ~ 123 Directions
124 ~ 135 Dragons

Format:
Haipai:
hai0 = "[[Hais]]"

Tsumos:
<T[Hai]/>
<U[Hai]/>
<V[Hai]/>
<W[Hai]/>

Discards:
<D[Hai]/>
<E[Hai]/>
<F[Hai]/>
<G[Hai]/>
"""
from mahjong.shanten import Shanten
import numpy as np
import re
# File to Open should be mjlog
# File to Write should be csv file
class MjlogToCSV:
    def __init__(self, file_to_open, file_to_write):
        self.file_to_open = file_to_open
        self.file_to_write = file_to_write
        self.mjlog = open(self.file_to_open, "r")
        self.csv = open(self.file_to_write, "w")
        self.text = self.mjlog.read()

    def getTehais(self):
        hand = np.zeros((4,34))
        shanten = Shanten()
        discard = np.array(4)
        text = self.text
        #p1 = re.compile(r'hai\d="\d+"')
        initialHands = []
        """for i in range(4):
            Hand = re.findall('hai' + str(i) + '="(.+?)"',text)
            Hand = [kyoku.split(",") for kyoku in Hand]
            Hand = [[self.haiConverter(int(tile)) for tile in kyoku] for kyoku in Hand]
            initialHands.append(Hand)
            """
        inits = []
        for val in (re.finditer("<INIT", text)):
            inits.append(val.span())

        for i in range(len(inits)):
            csvfile = open("csvs/data%d.csv"%(i),"w")
            hand = np.zeros((4,34))
            discards = []
            if(i < len(inits) -1):
                text = self.text[inits[i][1]:inits[i+1][0]]
            else:
                text = self.text[inits[i][1]:]
            for j in range(4):
                Hand = re.findall('hai' + str(j) + '="(.+?)"',text)
                Hand = [kyoku.split(",") for kyoku in Hand]
                Hand = [[self.haiConverter(int(tile)) for tile in kyoku] for kyoku in Hand]
                for k in range(len(Hand[0])):
                    hand[j][Hand[0][k]] += 1

            p1Tsumo = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<T\d+',text)]
            p2Tsumo = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<U\d+',text)]
            p3Tsumo = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<V\d+',text)]
            p4Tsumo = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<W\d+',text)]
            p1Discards = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<D\d+',text)]
            p2Discards = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<E\d+',text)]
            p3Discards = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<F\d+',text)]
            p4Discards = [self.haiConverter(int(tile[2:])) for tile in re.findall(r'<G\d+',text)]
            Tsumos = [p1Tsumo, p2Tsumo, p3Tsumo, p4Tsumo]
            Discards = [p1Discards, p2Discards, p3Discards, p4Discards]

            for player in range(4):
                discard = []
                smaller = len(Tsumos[player])
                if(len(Discards[player]) < len(Tsumos[player])):
                    smaller = len(Discards[player])
                for k in range(smaller):
                    hand[player][Tsumos[player][k]] += 1
                    hand[player][Discards[player][k]] -= 1
                    print(hand)
                    discard.append(Discards[player][k])
                    target = shanten.calculate_shanten(hand[player])
                discards.append(discard)


    def haiConverter(self, tile):
        tile = tile / 4
        return tile

def main():
    mj = MjlogToCSV("mjlogs/sample2.mjlog", "test.csv")
    mj.getTehais()


main()
